src/compiler/parser.ts

```diff
@@ -3569,18 +3569,51 @@ namespace ts {
             return parsePostfixTypeOrHigher();
         }
 
+        function parseFunctionOrConstructorTypeToError(
+            isInUnionType: boolean
+        ): TypeNode | undefined {
+            // the function type and constructor type shorthand notation
+            // are not allowed directly in unions and intersections, but we'll
+            // try to parse them gracefully and issue a helpful message.
+            if (isStartOfFunctionType() || token() === SyntaxKind.NewKeyword) {

```

Comment:
is there an already-defined `isStartOfConstructorType`? I'll go check...

Reply:
nope, but the only other call *also* is followed by `|| token() === SntaxKind.NewKeyword`. Can you move that clause into the function?